src/CommandLine.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "CommandLine.h"
+
+static void
+ResetGetopt(const bool allowStderrWarnings)
+{
+    opterr = allowStderrWarnings ? 1 : 0;
+    // getopt(3) uses global state but resets it if optind is zero
+    // getopt(3) always skips argv[0], even if optind is zero
+    optind = 0;
```

Comment:
Why optind is 0? From the getopt_long() man page:

```
The variable optind is the index of the next element to be processed in argv.
The system initializes this value to  1. The  caller can reset it to 1 to restart
scanning of the same argv, or when scanning a new argument vector.
```

Reply:
getopt uses a _getopt_data global to keep search state between getopt() calls. That state includes pointers and offsets. For example, the _getopt_data::__nextchar pointer is often reused between calls. Different CommandLine objects cannot share the same search state so we have to initialize. getopt(3) documentation mentions this, but in an ambiguous (or grammatically incorrect) way:

```
       A program that scans multiple argument vectors,  or  rescans  the  same
       vector  more than once, and [...X...],  or  [...Y...],
       must reinitialize getopt() by resetting
       optind to 0, rather than the traditional value of 1.  (Resetting  to  0
       forces  the  invocation  of  an  internal  initialization  routine that
       [...does Z..].)
```

None of the X, Y, and Z conditions/effects apply to us (today), but common sense, my reading of the getopt source [code](https://stackoverflow.com/questions/48278881/gdb-complaining-about-missing-raise-c/48287761#48287761), and my (dirty) tests seem to confirm the necessity of reseting the global state between independent getopt() calls.

Please rephrase the comment to avoid similar future questions.

Reply:
FYI: I made a simple experiment, parsing the same argv twice, resetting optind to '0' or '1' after the first pass, and got identical(correct) results.

Reply:
You cannot disprove my assertion that zero value is generally necessary by citing cases where it is not.

Reply:
Done with the comment.